imports/client/hooks/useCallState.ts

```diff
@@ -0,0 +1,805 @@
+        iceCandidates: JSON.parse(iceCandidates),
+        dtlsParameters: JSON.parse(serverDtlsParameters),
+        appData: {
+          _id,
+        },
+      });
+      newTransport.on('connect', ({ dtlsParameters: clientDtlsParameters }, callback) => {
+        connectRef.current = callback;
+        // No need to set a callback here, since the ConnectAck record acts as a
+        // callback
+        mediasoupConnectTransport.call({
+          transportId: _id,
+          dtlsParameters: JSON.stringify(clientDtlsParameters),
+        }, (err) => {
+          if (err) {
+            console.error(`Failed to connect transport ${direction}`, err);
+          }
+        });
+      });
+      log(`setting ${direction} transport`, newTransport);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: newTransport,
+      });
+      return () => {
+        if (!newTransport.closed) newTransport.close();
+      };
+    } else {
+      log(`clearing ${direction} transport`);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: undefined,
+      });
+      return undefined;
+    }
+  }, [
+    device, hasParams, transportParams?._id, direction,
+    transportParams?.transportId, transportParams?.iceParameters, transportParams?.iceCandidates,
+    transportParams?.dtlsParameters, dispatch,
+  ]);
+
+  return connectRef;
+};
+
+function cleanupProducerMapEntry(map: Map<string, ProducerState>, trackId: string) {
+  const producerState = map.get(trackId);
+  if (producerState) {
+    // Stop the producer if present
+    if (producerState.producer) {
+      log('stopping producer for track', trackId);
+      producerState.producer.close();
+    }
+
+    // Stop the producer sub if present.
+    if (producerState.subHandle) {
+      log('stopping producer sub for track', trackId);
+      producerState.subHandle.stop();
+    }
+
+    // Drop the removed track from the producerMapRef.
+    log('producerMapRef.delete', trackId);
+    map.delete(trackId);
+  }
+}
+
+type ProducerCallback = ({ id }: { id: string }) => void;
+type ProducerState = {
+  producer: types.Producer | undefined;
+  subHandle: Meteor.SubscriptionHandle | undefined;
+  producerServerCallback: ProducerCallback | undefined;
+  kind: string | undefined;
+  rtpParameters: string | undefined;
+}
+
+type ConsumerState = {
+  consumer: types.Consumer | undefined;
+  peerId: string;
+}
+
+const useCallState = ({ huntId, puzzleId, tabId }: {
+  huntId: string,
+  puzzleId: string,
+  tabId: string,
+}): [CallState, React.Dispatch<Action>] => {
+  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
+
+  useEffect(() => {
+    // If huntId, puzzleId, or tabId change (but mostly puzzleId), reset
+    // call state.
+    return () => {
+      log('huntId/puzzleId/tabId changed, resetting call state');
+      dispatch({ type: 'reset' });
+    };
+  }, [huntId, puzzleId, tabId]);
```

Comment:
I think it's a good idea, but the obvious implementation doesn't appear to do anything because we need to hook into react-router instead, since it's not a browser-level navigation that any of our links trigger.

This was something that was included in react-router v5, but is absent in v6: https://reactrouter.com/en/main/upgrading/v5#prompt-is-not-currently-supported .  Given that this looks like it'll be a bit more work to sort out, I filed #1037 for future followup.